### ISSUE TITLE
Kubelet restart cause running pod restart with UnexpectedAdmissionError when pods have initContainers and external devices like GPU 

### ISSUE THREAD

**Issue by zwpaper** [kubernetes/kubernetes#124345]
### What happened?

When restarting kubelet, it will restart the running pods with UnexpectedAdmissionError when pods' initContainers and containers both use external devices like GPU 

### What did you expect to happen?

Restart kubelet should not cause running pods to restart

### How can we reproduce it (as minimally and precisely as possible)?

1. create some pods with nvidia.com/gpu and initContainers which also use GPUs
2. wait until initContainers exit
3. restart kubelet

### Anything else we need to know?

there was an issue and a fix for running pods with devices, but it looks like the initContainers is not counted as should skip containers.

the fix has cherry-picked to v1.25.16

- issue: https://github.com/kubernetes/kubernetes/issues/118559
- fix: https://github.com/kubernetes/kubernetes/pull/118635
- cherry-pick v1.25 since v1.25.14: https://github.com/kubernetes/kubernetes/pull/119707

Related Issues:
- Completed Pod will also be affected: https://github.com/kubernetes/kubernetes/issues/117955
- memory manager related kubelet restart cause UnexpectedAdmissionError: https://github.com/kubernetes/kubernetes/issues/123971

### Kubernetes version

<details>

```console
$ kubectl version
# 1.25.16
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by ffromani** [comment]
my 2c: completed containers, aka containers terminated succesfully and not deleted, like init containers or containers belonging to Jobs, should NOT be restarted or even retry admission when kubelet is restarted. The reason is, well, these containers already completed succesfully.

**Comment by ffromani** [comment]
quoting https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#detailed-behavior:
```
A Pod cannot be Ready until all init containers have succeeded. The ports on an init container are not aggregated under a Service. A Pod that is initializing is in the Pending state but should have a condition Initialized set to false.

If the Pod [restarts](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#pod-restart-reasons), or is restarted, all init containers must execute again.
```
and
```
Because init containers can be restarted, retried, or re-executed, init container code should be idempotent. In particular, code that writes to files on EmptyDirs should be prepared for the possibility that an output file already exists.
```
which ties to https://github.com/kubernetes/kubernetes/issues/123980

**Comment by ffromani** [comment]
xref: https://github.com/kubernetes/kubernetes/issues/117955

**Comment by zwpaper** [comment]
> my 2c: completed containers, aka containers terminated succesfully and not deleted, like init containers or containers belonging to Jobs, should NOT be restarted or even retry admission when kubelet is restarted. The reason is, well, these containers already completed succesfully.

yes! that's why I created this issue, and this issue is more focus on the init-containers, the completed init-containers cause running pods to be restarted.

this fixed doesn't help because the pod is still running: https://github.com/kubernetes/kubernetes/commit/a2ca66d280df83c5d32e058ef97a3d117a8309ed

I was thinking about, can we add a check here to skip allocateDevice if containers is expected stopped, looks like my pods was killed here:

https://github.com/ffromani/kubernetes/blob/7e3638982acebe901b34f3d9bab4f4e4c5d703c9/pkg/kubelet/cm/devicemanager/manager.go#L811

**Comment by AnishShah** [comment]
@zwpaper can you try this on a newer k8s version? 1.25 is out of support. Can you also share a pod spec to reproduce this error? Why is the pod failing with an admission error if the device is still available?

**Comment by zwpaper** [comment]
> @zwpaper can you try this on a newer k8s version? 

Of cause, but maybe need some another time, once done, i will update it here.
But we have checked the kubelet logic and added some logs to make sure the position, and double checked the master branch is still not changed, so that it should be easy to reproduce.


> Why is the pod failing with an admission error if the device is still available?

The is a bug related in nv device plugins that may crash when kubelet restarts, but i have not yet deep dived into it.
But even without that bug, a kubelet restart will ask all the device plugins to restart and register their socks by design, so that there will be a chance there will be no healthy devices when kubelet starts.
The main problem is that kubelet should not allocate devices for the completed containers like completed jobs and init containers

**Comment by nareshku** [comment]
I also encountered this issue. From my observations, only the Pods that had **completed** and previously **failed** were retried, and they failed again with the error shown below.

```
I0129 05:42:59.060605 3736781 event.go:294] "Event occurred" object="<namespace>/<podname>" kind="Pod" apiVersion="v1" type="Warning" reason="UnexpectedAdmissionError" message="Allocate failed due to requested number of devices unavailable for nvidia.com/gpu. Requested: 4, Available: 0, which is unexpected"

I0129 05:42:59.062040 3736781 kubelet.go:2155] "SyncLoop (PLEG): event for pod" pod="<namespace>/<podname>" event=&{ID:.... Type:ContainerDied Data:....}
```
What's the workaround for this other than safely draining and restarting Kubelet?

**Comment by ffromani** [comment]
/triage accepted

there's work to do in this area.
In recent (kube >= 1.32 if not 1.30, need to check) failed pods should not be retry admission again, though.
@nareshku which kube version were you running?

**Comment by nareshku** [comment]
> /triage accepted
> 
> there's work to do in this area. In recent (kube >= 1.32 if not 1.30, need to check) failed pods should not be retry admission again, though. [@nareshku](https://github.com/nareshku) which kube version were you running?

kube version v1.25.

Also, can you please confirm if the fix #126343 address this issue?

**Comment by ffromani** [comment]
> Also, can you please confirm if the fix [#126343](https://github.com/kubernetes/kubernetes/pull/126343) addresses this issue?

This is the change I was hinting at (and which I didn't have handy, yet). There could be further work to be done in this area though, which we need to check. This is what I meant in https://github.com/kubernetes/kubernetes/issues/124345#issuecomment-2685931651

**Comment by nareshku** [comment]
@ffromani We added the fix from https://github.com/kubernetes/kubernetes/pull/126343 and built Kubelet locally and tested it. The fix worked and address this issue.
I am wondering if it is feasible to cherry pick this change and include it as part of previous releases?
